cli/module_graph2.rs

```diff
@@ -845,6 +835,155 @@ impl Graph2 {
     self.modules.contains_key(s)
   }
 
+  pub fn emit(self, options: EmitOptions) -> EmitResult {
+    let mut config = TsConfig::new(json!({
+      "allowJs": true,
+      // TODO(@kitsonk) consider enabling this by default
+      //   see: https://github.com/denoland/deno/issues/7732
+      "emitDecoratorMetadata": false,
+      "esModuleInterop": true,
+      "experimentalDecorators": true,
+      "isolatedModules": true,
+      "jsx": "react",
+      "lib": TypeLib::DenoWindow,
+      "module": "esnext",
+      "strict": true,
+      "target": "esnext",
+    }));
+    let opts = match options.bundle_type {
+      BundleType::Esm => json!({
+        "checkJs": false,
+        "inlineSourceMap": false,
+        "noEmit": true,
+        "jsxFactory": "React.createElement",
+        "jsxFragmentFactory": "React.Fragment",
+      }),
+      BundleType::None => json!({
+        "outDir": "deno://",
+        "removeComments": true,
+        "sourceMap": true,
+      }),
+    };
+    config.merge(&opts);
+    let maybe_ignored_options =
+      if let Some(user_options) = &options.maybe_user_config {
+        config.merge_user_config(user_options)?
+      } else {
+        None
+      };
+
+    let root_names = self.get_root_names();
+    let hash_data =
+      vec![config.as_bytes(), version::DENO.as_bytes().to_owned()];
+    let graph = Rc::new(RefCell::new(self));
+
+    let response = exec(
+      js::compiler_isolate_init(),
+      Request {
+        config: config.clone(),
+        debug: options.debug,
+        graph: graph.clone(),
+        hash_data,
+        maybe_tsbuildinfo: None,
+        root_names,
+      },
+    )?;
+
+    let mut emitted_files = HashMap::new();
+    match options.bundle_type {
+      BundleType::Esm => {
+        assert!(
+          response.emitted_files.is_empty(),
+          "No files should have been emitted from tsc."
+        );
+        let graph = graph.borrow();
+        assert_eq!(
+          graph.roots.len(),
+          1,
+          "Only a single root module supported."
+        );
+        let specifier = &graph.roots[0];
+        let s = graph.emit_bundle(specifier, &config.into())?;
+        emitted_files.insert("deno:///bundle.js".to_string(), s);
+      }
+      BundleType::None => {
+        for emitted_file in &response.emitted_files {
+          assert!(
+            emitted_file.maybe_specifiers.is_some(),
+            "Orphaned file emitted."
+          );
+          let specifiers = emitted_file.maybe_specifiers.clone().unwrap();
+          assert_eq!(
+            specifiers.len(),
+            1,
+            "An unexpected number of specifiers associated with emitted file."
+          );
+          let specifier = specifiers[0].clone();
+          let extension = match emitted_file.media_type {
+            MediaType::JavaScript => ".js",
+            MediaType::SourceMap => ".js.map",
+            _ => unreachable!(),
+          };
```

Comment:
Won't this bit cause double extensions for specifiers with extensions?

Eg. `file://foo/bar/bazz.js` -> `file://foo/bar/bazz.js.js`, `file://foo/bar/bazz.js.js.map`

Reply:
Yeah, but it is consistent... because if we replace the extension, we could end up in situations where the input specifier == emit specifier, and if you as a user are using local specifiers and then taking the emit and writing them out to disk yourself, you might overwrite your sources.  I would prefer to make it clear in the documentation that this occurs versus running the risk of someone overwriting their input.

Reply:
That sounds good to me 👍